pennylane/transforms/optimization/merge_rotations.py

```diff
@@ -0,0 +1,128 @@
+    2: ──╰X──────H──────────────╰C──────┤
+
+    By inspection, we can combine the two ``RX`` rotations on the first qubit.
+    On the second qubit, we have a cumulative angle of 0, and the gates will cancel.
+
+    >>> optimized_qfunc = merge_rotations(qfunc)
+    >>> optimized_qnode = qml.QNode(optimized_qfunc, dev)
+    >>> print(qml.draw(optimized_qnode)(1, 2, 3))
+    0: ───RX(3)─────╭RZ(3)──┤ ⟨Z⟩
+    1: ──╭C─────────│───────┤
+    2: ──╰X──────H──╰C──────┤
+    """
+    # Make a working copy of the list to traverse
+    list_copy = tape.operations.copy()
+
+    while len(list_copy) > 0:
+        current_gate = list_copy[0]
+
+        # Normally queue any non-rotation gates
+        if not current_gate.is_composable_rotation:
+            apply(current_gate)
+            list_copy.pop(0)
+            continue
+
+        # Otherwise, find the next gate that acts on the same wires
+        next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        # If no such gate is found (either there simply is none, or there are other gates
+        # "in the way", queue the operation and move on
+        if next_gate_idx is None:
+            apply(current_gate)
+            list_copy.pop(0)
+            continue
+
+        # We need to use stack to get this to work and be differentiable in all interfaces
+        cumulative_angles = stack(current_gate.parameters)
+
+        # As long as there is a valid next gate, check if we can merge the angles
+        while next_gate_idx is not None:
+            # Get the next gate
+            next_gate = list_copy[next_gate_idx + 1]
+
+            # If next gate is of the same type, we can merge the angles
+            if current_gate.name == next_gate.name and current_gate.wires == next_gate.wires:
+                list_copy.pop(next_gate_idx + 1)
+
+                # The Rot gate must be treated separately
+                if current_gate.name == "Rot":
+                    cumulative_angles = _fuse_rot_angles(
+                        cumulative_angles, cast_like(stack(next_gate.parameters), cumulative_angles)
+                    )
+                # Other, single-parameter rotation gates just have the angle summed
+                else:
+                    cumulative_angles = cumulative_angles + cast_like(
+                        stack(next_gate.parameters), cumulative_angles
+                    )
+            # If it is not, we need to stop
+            else:
+                break
+
+            # If we did merge, look now at the next gate
+            next_gate_idx = _find_next_gate(current_gate.wires, list_copy[1:])
+
+        if not allclose(cumulative_angles, zeros(len(cumulative_angles))):
+            current_gate.__class__(*cumulative_angles, wires=current_gate.wires)
+
+        # Remove the first gate gate from the working list
+        list_copy.pop(0)
+
+    # Queue the measurements normally
+    for m in tape.measurements:
+        apply(m)
```

Comment:
It strikes me that so much of this for loop is similar to the previous transform... I wonder if there is a way of 'combining' the for loops if both are applied consecutively 🤔

Reply:
Oh another question - I suppose this transform would have to be run multiple times to merge all Rot gates?

Reply:
It should actually glob together arbitrary-length sequences, so if there 3 `Rot` gates in a row, it will end up as a single `Rot` in just one pass. Probably a good thing to add to a test!

Reply:
> It strikes me that so much of this for loop is similar to the previous transform... I wonder if there is a way of 'combining' the for loops if both are applied consecutively

We *could* just combine them into a single transform, but I kind of like the flexibility of them being separate. Also, the way the loops are set up, you might run into problems with circuits like the following:
```
qml.RZ(0.3, wires=0)
qml.Hadamard(wires=0)
qml.Hadamard(wires=0)
qml.RZ(0.2, wires=0)
```
If we do only one pass-through of a combined version, the two Hadamards would cancel, but the two `RZ` wouldn't merge, because we've moved ahead to the second `RZ` and it won't see the one behind. To catch them both, we'd need a second pass, at which point we may as well have a separate transform.

Reply:
got it :+1: